standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/metrics/AcidMetricService.java

```diff
@@ -0,0 +1,157 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.metastore.metrics;
+
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.metastore.MetaStoreThread;
+import org.apache.hadoop.hive.metastore.api.MetaException;
+import org.apache.hadoop.hive.metastore.api.ShowCompactRequest;
+import org.apache.hadoop.hive.metastore.api.ShowCompactResponse;
+import org.apache.hadoop.hive.metastore.api.ShowCompactResponseElement;
+import org.apache.hadoop.hive.metastore.conf.MetastoreConf;
+import org.apache.hadoop.hive.metastore.txn.TxnStore;
+import org.apache.hadoop.hive.metastore.txn.TxnUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+
+/**
+ * Collect and publish ACID and compaction related metrics.
+ */
+public class AcidMetricService extends Thread implements MetaStoreThread {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AcidMetricService.class);
+  private Configuration conf;
+  private int threadId;
+  private AtomicBoolean stop;
+  private TxnStore txnHandler;
+  private long checkInterval;
+
+  @Override
+  public void setThreadId(int threadId) {
+    setPriority(MIN_PRIORITY);
+    this.threadId = threadId;
+  }
+
+  @Override
+  public void init(AtomicBoolean stop) throws Exception {
+    setDaemon(true); // the process will exit without waiting for this thread
+
+    this.stop = stop;
+    checkInterval = MetastoreConf.getTimeVar(conf, MetastoreConf.ConfVars.METASTORE_ACIDMETRICS_CHECK_INTERVAL, TimeUnit.MILLISECONDS);
+    // Get our own instance of the transaction handler
+    txnHandler = TxnUtils.getTxnStore(conf);
+  }
+
+  @Override
+  public void run() {
+    LOG.info("Starting AcidMetricService thread");
+    try {
+      do {
+        long startedAt = System.currentTimeMillis();
+
+        boolean metricsEnabled = MetastoreConf.getBoolVar(conf, MetastoreConf.ConfVars.METRICS_ENABLED) &&
+            MetastoreConf.getBoolVar(conf, MetastoreConf.ConfVars.METASTORE_ACIDMETRICS_THREAD_ON);
+        if (!metricsEnabled) {
+          // Make it possible to disable metrics collection without a restart
+          LOG.debug("AcidMetricService is running but metric collection is not enabled");
+          if (!stop.get()) {
+            Thread.sleep(checkInterval);
+          }
+          continue;
+        }
+        try {
+          collectMetrics();
+        } catch (Exception ex) {
+         LOG.error("Caught exception in AcidMetricService loop", ex);
+        }
+
+        long elapsedTime = System.currentTimeMillis() - startedAt;
+        LOG.info("AcidMetricService thread finished one loop in {} seconds.", elapsedTime / 1000);
```

Comment:
I vote for debug level here (yes this is hypocritical :) )

Reply:
done